File: Java/Sorting/TreeSort.java
/*Approach:
-Tree sort,also known as binary tree sort,is a sorting algorithm that builds a binary search tree from the elements to be sorted and then performs an in-order traversal on the tree to retrieve the elements in sorted order.
-Define a private inner class Node to represent a node in the binary search tree with an integer value, height, left, and right references.
-Implement the height() method to calculate the height of a node, returning -1 for a null node.
-Implement the isEmpty() method to check if the tree is empty (root is null).
-Implement the populate(int[] arr) method to build the binary search tree by inserting each element from the input array using the insert method.
-Implement the insert(int value) method to insert a new integer value into the binary search tree recursively, updating the node heights for balancing.
-Implement the inOrder() method to perform an in-order traversal and print the values of the nodes in ascending order.

Note:The time complexity and space complexity of the Tree Sort approach depends on how the binary search tree looks like,which is determined by whether the tree is balanced or unbalanced.

TimeComplexity:
    -Best-case time complexity(when the tree is highly balanced):O(n log n)
    -Average-case time complexity:O(n log n)
    -Worst-case time complexity(when the tree be comes a linear chain):O(n^2)
SpaceComplexity:
    -Best-case space complexity (balanced tree): O(log n)
    -Average-case space complexity: O(log n) to O(n) (depending on the balance)
    -Worst-case space complexity (unbalanced tree): O(n)

SampleInput:
Array before sorting: 15 10 5 8 12 20

SampleOutput:
Array after sorting: 5 8 10 12 15 20
 */

public class TreeSort {
    private class Node{
        int value;
        int height;
        Node left;
        Node right;

        public Node(int value){
            this.value=value;
        }

        public int getValue() {
            return value;
        }
    }

    private Node root;

    public int height(Node node){
        if(node==null){
            return -1;
        }
        return node.height;
    }

    public boolean isEmpty() {
        return root == null;
    }

    public void populate(int[] arr){
        for(int i=0;i< arr.length;i++){
            this.insert(arr[i]);
        }
    }
    //insertion of the element in binary tree search format
    //smaller elements go to the left subtree,and larger elements go to the right subtree.If an element is already present in the BST(duplicate value),
    // it is usually placed in the right subtree.
    public void insert(int value){
        root=insert(value,root);
    }

    private Node insert(int value, Node node){
        if(node==null){
            node=new Node(value);
            return node;
        }
        if(value<node.value){
            node.left=insert(value,node.left);
        }
        if(value>node.value)
        {
            node.right=insert(value,node.right);
        }

        node.height = Math.max(height(node.left), height(node.right)) + 1;
        return node;
    }
    //in-order traversal,the left subtree is explored first,then the current node,and finally,the right subtree.
    public void inOrder(){
        inOrder(root);
    }

    private void inOrder(Node node){
        if(node==null){
            return;
        }
        inOrder(node.left);
        System.out.println(node.value);
        inOrder(node.right);
    }

    public static void main(String[] args){
        // Sample input
        int arr[]={15,10,5,8,12,20};
        TreeSort sort=new TreeSort();
        sort.populate(arr);
        sort.inOrder();
    }
}

File: Python/Sorting/TreeSort.py
"""
Approach:
- Tree sort is a sorting algorithm that builds a binary search tree from the elements to be sorted and
  then performs an in-order traversal of the tree to retrieve the elements in sorted order.

Time Complexity:
- The time complexity of the Tree Sort algorithm depends on how balanced the binary search tree is.
  In the worst case, when the tree is highly unbalanced and resembles a linked list, the time complexity of Tree Sort
  is O(n^2), but in the average and best cases (when the tree is approximately balanced), it's O(n * log(n)).

Space Complexity:
- The space complexity of the Tree Sort algorithm primarily depends on the space required for the
  binary search tree (BST) and any additional data structures used for recursion. Here's a breakdown of the space
  complexity components:

  Binary Search Tree (BST): The space required for the binary search tree is determined by the number of nodes in the
  tree. In the worst case (highly unbalanced tree), the tree can have up to n nodes, where n is the number of
  elements in the input array. Therefore, in the worst case, the space complexity for the BST is O(n).

  Recursion Stack: During the in-order traversal of the BST, the algorithm uses recursion to visit nodes. The space
  complexity of the recursion stack is determined by the maximum depth of the recursion. In the worst case,
  when the tree is highly unbalanced, the recursion stack can have a depth of n (all elements in the tree are in a
  single path). In the average and best cases (balanced tree), the depth of recursion is limited to O(log(n)).
  Therefore, in the worst case, the space complexity for the recursion stack is O(n), and in the average/best cases,
  it's O(log(n)).
  Overall, the dominant factor in the space complexity of Tree Sort is typically the space required
  for the binary search tree. So, the space complexity is O(n) in the worst case and O(log(n)) in the average and
  best cases, considering both the BST and recursion stack.

Sample Input:
- arr = [5, 2, 9, 1, 5, 6]

Sample Output:
- Sorted array: [1, 2, 5, 5, 6, 9]
"""


class TreeNode:
    def __init__(self, key):
        self.left = None
        self.right = None
        self.val = key


def insert(root, key):
    # Function to insert a new element into the binary search tree.
    if root is None:
        return TreeNode(key)
    else:
        if root.val < key:
            root.right = insert(root.right, key)
        else:
            root.left = insert(root.left, key)
    return root


def in_order_traversal(root):
    # Function to perform an in-order traversal of the binary search tree.
    result = []
    if root:
        result = in_order_traversal(root.left)
        result.append(root.val)
        result = result + in_order_traversal(root.right)
    return result


def tree_sort(array):
    # Function to sort an array using Tree Sort.
    root = None
    for item in array:
        root = insert(root, item)  # Insert each element into the BST.
    sorted_array = in_order_traversal(root)  # Traverse the BST in-order to get the sorted elements.
    return sorted_array


if __name__ == "__main__":
    # Example usage:
    arr = [5, 2, 9, 1, 5, 6]
    sorted_arr = tree_sort(arr)
    print("Sorted array:", sorted_arr)
